// firebase-tools/lib/emulator/functionsEmulatorShell.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FunctionsEmulatorShell = void 0;
const uuid = require("uuid");
const functionsEmulator_1 = require("./functionsEmulator");
const functionsEmulatorShared_1 = require("./functionsEmulatorShared");
const utils = require("../utils");
const logger_1 = require("../logger");
const error_1 = require("../error");
class FunctionsEmulatorShell {
    constructor(emu) {
        this.emu = emu;
        this.urls = {};
        this.triggers = emu.getTriggerDefinitions();
        this.emulatedFunctions = this.triggers.map((t) => t.id);
        const entryPoints = this.triggers.map((t) => t.entryPoint);
        utils.logLabeledBullet("functions", `Loaded functions: ${entryPoints.join(", ")}`);
        for (const trigger of this.triggers) {
            if (trigger.httpsTrigger) {
                this.urls[trigger.id] = functionsEmulator_1.FunctionsEmulator.getHttpFunctionUrl(this.emu.getInfo().host, this.emu.getInfo().port, this.emu.getProjectId(), trigger.name, trigger.region);
            }
        }
    }
    call(id, data, opts) {
        const trigger = this.getTrigger(id);
        logger_1.logger.debug(`shell:${id}: trigger=${JSON.stringify(trigger)}`);
        logger_1.logger.debug(`shell:${id}: opts=${JSON.stringify(opts)}, data=${JSON.stringify(data)}`);
        if (!trigger.eventTrigger) {
            throw new error_1.FirebaseError(`Function ${id} is not a background function`);
        }
        const eventType = trigger.eventTrigger.eventType;
        let resource = opts.resource;
        if (typeof resource === "object" && resource.name) {
            resource = resource.name;
        }
        const proto = {
            eventId: uuid.v4(),
            timestamp: new Date().toISOString(),
            eventType,
            resource,
            params: opts.params,
            auth: opts.auth,
            data,
        };
        this.emu.startFunctionRuntime(id, trigger.name, functionsEmulatorShared_1.EmulatedTriggerType.BACKGROUND, proto);
    }
    getTrigger(id) {
        const result = this.triggers.find((trigger) => {
            return trigger.id === id;
        });
        if (!result) {
            throw new error_1.FirebaseError(`Could not find trigger ${id}`);
        }
        return result;
    }
}
exports.FunctionsEmulatorShell = FunctionsEmulatorShell;
